c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+            hir_id: ex.hir_id,
+            span: ex.span,
+            desc,
+        };
+        match self.mir_map.entry((loc, sub_loc.to_owned())) {
+            Entry::Vacant(e) => {
+                e.insert(origin);
+            }
+            Entry::Occupied(e) => {
+                if *e.get() != origin {
+                    let old_origin = e.get().clone();
+                    error!(
+                        "conflicting origins for {:?} {:?} ({:?})\n\
+                            origin 1 = {:?}\n\
+                            origin 2 = {:?}\n\
+                            expr 2 = {:?}",
+                        loc,
+                        sub_loc,
+                        self.location_span(loc),
+                        origin,
+                        old_origin,
+                        ex,
+                    );
+                }
+            }
+        }
+    }
+
+    fn get_sole_assign(
+        &self,
+        locs: &[Location],
+    ) -> Option<(Location, mir::Place<'tcx>, &'a mir::Rvalue<'tcx>)> {
+        if locs.len() != 1 {
+            return None;
+        }
+        self.get_last_assign(locs)
+    }
+
+    fn get_last_assign(
+        &self,
+        locs: &[Location],
+    ) -> Option<(Location, mir::Place<'tcx>, &'a mir::Rvalue<'tcx>)> {
+        let &loc = locs.last()?;
+        let stmt = self.mir.stmt_at(loc).left()?;
+        match stmt.kind {
+            mir::StatementKind::Assign(ref x) => Some((loc, x.0, &x.1)),
+            _ => None,
+        }
+    }
+
+    fn get_last_call(
+        &self,
+        locs: &[Location],
+    ) -> Option<(
+        Location,
+        mir::Place<'tcx>,
+        &'a mir::Operand<'tcx>,
+        &'a [mir::Operand<'tcx>],
+    )> {
+        let &loc = locs.last()?;
+        let term = self.mir.stmt_at(loc).right()?;
+        match term.kind {
+            mir::TerminatorKind::Call {
+                ref func,
+                ref args,
+                destination,
+                ..
+            } => Some((loc, destination, func, args)),
+            _ => None,
+        }
+    }
+
+    fn should_ignore_statement(&self, loc: Location) -> bool {
+        if let Some(stmt) = self.mir.stmt_at(loc).left() {
+            match stmt.kind {
+                mir::StatementKind::FakeRead(..)
+                | mir::StatementKind::StorageLive(..)
+                | mir::StatementKind::StorageDead(..)
+                | mir::StatementKind::Nop => return true,
+                _ => {}
+            }
+        }
+        false
+    }
+
+    fn is_var(&self, pl: mir::Place<'tcx>) -> bool {
+        pl.projection.len() == 0
+    }
+
+    fn visit_expr_inner(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        let _g = panic_detail::set_current_span(ex.span);
+
+        let mut locs = Vec::with_capacity(1);
+        for &loc in self.span_index.lookup_exact(ex.span) {
+            if !self.should_ignore_statement(loc) {
+                locs.push(loc);
+            }
+        }
```

Comment:
I don't think this is any clearer than the original.  This foreach/if/push pattern is extremely common - I think most developers have seen code like this hundreds of times in dozens of different languages and would recognize the pattern instantly.

Reply:
Since we're using `Vec::new()` now, this can be written like this instead:

```suggestion
        let locs = self
            .span_index
            .lookup_exact(ex.span)
            .copied()
            .filter(|&loc| !self.should_ignore_statement(loc))
            .collect::<Vec<_>>();
```

From what I've seen, that's a lot more common and idiomatic in Rust.  `locs` doesn't have to be `mut` anymore, and `.collect()` can take advantage of the `.size_hint()` that `.push()` can't.

Reply:
That looks absolutely horrible (`rustfmt` tends to make a mess of these kinds of iterator chains), but you're right that `collect` is the thing to use here.  I didn't check the context and assumed there were more `push`es elsewhere in the function.